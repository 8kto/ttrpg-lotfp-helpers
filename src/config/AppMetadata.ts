/* istanbul ignore file */
import type { Metadata } from 'next'

// TODO translate
export const AppMetadata: Metadata = {
  description: 'Lamentations of the Flame Princess helpers',
  title: 'Princess Helpers',
}
