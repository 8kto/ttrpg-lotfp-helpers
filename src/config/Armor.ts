import { msg } from '@lingui/macro'

import type { EquipmentItemTranslated } from '@/config/types'
import type { ArmorItem } from '@/domain/armor'
import { ArmorType } from '@/domain/armor'
import { EncumbrancePoint } from '@/domain/encumbrance'

export const Armor: ReadonlyArray<EquipmentItemTranslated<ArmorItem>> = [
  {
    armorClass: 14,
    cityCost: 25,
    name: msg`Leather`,
    points: EncumbrancePoint.None,
    ruralCost: 50,
    type: ArmorType.Armor,
  },
  {
    armorClass: 16,
    cityCost: 100,
    name: msg`Chain`,
    points: EncumbrancePoint.Oversized,
    ruralCost: null,
    type: ArmorType.Armor,
  },
  {
    armorClass: 18,
    cityCost: 1000,
    details: msg`config.armor.plate.details`,
    name: msg`Plate`,
    points: EncumbrancePoint.Heavy,
    ruralCost: null,
    type: ArmorType.Armor,
  },
  {
    armorClass: '+1,+2',
    cityCost: 10,
    details: msg`+1 AC versus mêlée attacks, +2 AC versus missile attacks.`,
    name: msg`Shield`,
    points: EncumbrancePoint.Oversized,
    ruralCost: 25,
    type: ArmorType.Shield,
  },
  {
    armorClass: 14,
    cityCost: 250,
    name: msg`Leather Barding`,
    points: EncumbrancePoint.None,
    ruralCost: null,
    type: ArmorType.Barding,
  },
  {
    armorClass: 16,
    cityCost: 500,
    name: msg`Chain Barding`,
    points: EncumbrancePoint.Oversized,
    ruralCost: null,
    type: ArmorType.Barding,
  },
  {
    armorClass: 18,
    cityCost: 1000,
    name: msg`Plate Barding`,
    points: EncumbrancePoint.Heavy,
    ruralCost: null,
    type: ArmorType.Barding,
  },
]
