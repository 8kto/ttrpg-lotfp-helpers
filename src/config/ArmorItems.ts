import { t } from '@lingui/macro'

import type { ArmorItem } from '@/domain/armor'
import { ArmorType } from '@/domain/armor'
import { EncumbrancePoint } from '@/domain/encumbrance'

export const ArmorItems = (): ReadonlyArray<ArmorItem> => [
  {
    armorClass: 14,
    categoryKey: 'armor',
    cityCost: 25,
    name: t`Leather`,
    points: EncumbrancePoint.None,
    ruralCost: 50,
    type: ArmorType.Armor,
  },
  {
    armorClass: 16,
    categoryKey: 'armor',
    cityCost: 100,
    name: t`Chain`,
    points: EncumbrancePoint.Oversized,
    ruralCost: null,
    type: ArmorType.Armor,
  },
  {
    armorClass: 18,
    categoryKey: 'armor',
    cityCost: 1000,
    details: t`config.armor.plate.details`,
    name: t`Plate`,
    points: EncumbrancePoint.Heavy,
    ruralCost: null,
    type: ArmorType.Armor,
  },
  {
    armorClass: '+1,+2',
    categoryKey: 'armor',
    cityCost: 10,
    details: t`+1 AC versus mêlée attacks, +2 AC versus missile attacks.`,
    name: t`Shield`,
    points: EncumbrancePoint.Oversized,
    ruralCost: 25,
    type: ArmorType.Shield,
  },
  {
    armorClass: 14,
    categoryKey: 'armor',
    cityCost: 250,
    name: t`Leather Barding`,
    points: EncumbrancePoint.None,
    ruralCost: null,
    type: ArmorType.Barding,
  },
  {
    armorClass: 16,
    categoryKey: 'armor',
    cityCost: 500,
    name: t`Chain Barding`,
    points: EncumbrancePoint.Oversized,
    ruralCost: null,
    type: ArmorType.Barding,
  },
  {
    armorClass: 18,
    categoryKey: 'armor',
    cityCost: 1000,
    name: t`Plate Barding`,
    points: EncumbrancePoint.Heavy,
    ruralCost: null,
    type: ArmorType.Barding,
  },
]
