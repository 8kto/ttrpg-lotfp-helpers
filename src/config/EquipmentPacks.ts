import { msg } from '@lingui/macro'

import type { EquipmentPack } from '@/domain'

const Base: EquipmentPack = {
  items: [
    ['Bedroll', 1],
    ['Clothing, Normal', 1],
    ['Rations, Standard/Day', 3],
    ['Backpack', 1],
    ['Pouch', 1],
    ['Sack', 1],
    ['Tinderbox', 1],
  ],
  name: msg`Base items set`,
}

const AdventuringBasics: EquipmentPack = {
  items: [
    ['Chalk', 1],
    ['Grappling Hook', 1],
    [`Rope, 50'`, 1],
    ['Lantern', 1],
    ['Flask of Lamp Oil', 3],
    ['Tent, Personal', 1],
  ],
  name: msg`Adventuring Basics`,
}
const AdventurerReadyForAnything: EquipmentPack = {
  items: [
    [`Vial or Bottle`, 1],
    [`Spike, Iron `, 10],
    [`Pole, 10'`, 1],
    [`Paper'`, 10],
    [`Ink'`, 10],
    [`Scroll Case`, 1],
    [`Mirror, Steel`, 1],
  ],
  name: msg`Ready for anything`,
}

const AdventurerSeasoned: EquipmentPack = {
  items: [
    ['Book, Blank', 1],
    ['Caltrop', 10],
    ['Candle', 1],
    ['Chalk', 1],
    ['Crowbar', 1],
    ['Flask of Lamp Oil', 3],
    ['Ink', 1],
    ['Instrument', 1],
    ['Lantern', 1],
    ['Lock', 1],
    ['Mallet', 1],
    ['Mirror, Steel', 1],
    ['Nails', 20],
    ["Pole, 10'", 1],
    ["Rope, 50'", 1],
    ['Soap', 1],
    ['Spike, Wooden', 1],
    ['Tinderbox', 1],
    ['Torch', 3],
    ['Whistle', 1],
  ],
  name: msg`Seasoned adventurer`,
}

export const EquipmentPacks: Array<EquipmentPack> = [
  Base,
  AdventuringBasics,
  AdventurerReadyForAnything,
  AdventurerSeasoned,
]
