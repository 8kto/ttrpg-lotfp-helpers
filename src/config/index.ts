export { Armor } from './Armor'
export { MeleeWeapons } from './MeleeWeapons'
export { MiscEquipment } from './MiscEquipment'
export { MissileWeapons } from './MissileWeapons'
