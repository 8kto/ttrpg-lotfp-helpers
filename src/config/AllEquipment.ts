import type { EquipmentItem } from '@/domain'
import type { ArmorItem } from '@/domain/armor'
import type { MeleeWeaponItem, MissileWeaponItem } from '@/domain/weapon'
import deepclone from '@/shared/helpers/deepclone'

import { Armor } from './Armor'
import { MeleeWeapons } from './MeleeWeapons'
import { MiscEquipment } from './MiscEquipment'
import { MissileWeapons } from './MissileWeapons'

type Equipment = {
  MissileWeapons: ReadonlyArray<MissileWeaponItem>
  Armor: ReadonlyArray<ArmorItem>
  MeleeWeapons: ReadonlyArray<MeleeWeaponItem>
  MiscEquipment: ReadonlyArray<EquipmentItem>
}

export const AllEquipment: Readonly<Equipment> = deepclone({
  Armor,
  MeleeWeapons,
  MiscEquipment,
  MissileWeapons,
})
