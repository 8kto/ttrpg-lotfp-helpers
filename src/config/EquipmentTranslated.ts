import type { I18n } from '@lingui/core'

import type { EquipmentItemTranslated } from '@/config/types'
import type { EquipmentItem } from '@/domain'
import type { ArmorItem } from '@/domain/armor'
import type { MeleeWeaponItem, MissileWeaponItem } from '@/domain/weapon'
import deepclone from '@/shared/helpers/deepclone'
import type { EquipmentCategoryKey } from '@/state/InventoryState'

import { Armor } from './Armor'
import { MeleeWeapons } from './MeleeWeapons'
import { MiscEquipment } from './MiscEquipment'
import { MissileWeapons } from './MissileWeapons'

export default class EquipmentTranslated {
  private readonly trans: I18n['_']

  constructor(trans: I18n['_']) {
    this.trans = trans
  }

  private translate<T>(
    items: ReadonlyArray<EquipmentItemTranslated<T>>,
    categoryKey: EquipmentCategoryKey,
  ): ReadonlyArray<T> {
    return items.map((item) => {
      return {
        ...item,
        categoryKey,
        details: item.details ? this.trans(item.details) : undefined,
        name: typeof item.name === 'object' ? this.trans(item.name) : item.name,
      } as T
    })
  }

  get Armor(): ReadonlyArray<ArmorItem> {
    return this.translate<ArmorItem>(deepclone(Armor), 'armor')
  }

  get MeleeWeapons(): ReadonlyArray<MeleeWeaponItem> {
    return this.translate<MeleeWeaponItem>(
      deepclone(MeleeWeapons),
      'meleeWeapons',
    )
  }

  get MissileWeapons(): ReadonlyArray<MissileWeaponItem> {
    return this.translate<MissileWeaponItem>(
      deepclone(MissileWeapons),
      'missileWeapons',
    )
  }

  get MiscEquipment(): ReadonlyArray<EquipmentItem> {
    return this.translate<EquipmentItem>(
      deepclone(MiscEquipment),
      'miscEquipment',
    )
  }
}
