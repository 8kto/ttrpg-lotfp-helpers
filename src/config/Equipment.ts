import { Armor, MeleeWeapons, MiscEquipment, MissileWeapons } from '@/config'
import type { ArmorItem } from '@/domain/armor'
import type { EquipmentItem } from '@/domain/equipment'
import type { MeleeWeaponItem, MissileWeaponItem } from '@/domain/weapon'
// TODO use for other case
import deepclone from '@/shared/helpers/deepclone'

/**
 * Utilizes getters for each equipment category (Armor, MeleeWeapons, etc.)
 * to enable real-time translation when the locale changes.
 */
const Equipment = {
  get Armor(): ReadonlyArray<ArmorItem> {
    return deepclone(Armor())
  },

  get MeleeWeapons(): ReadonlyArray<MeleeWeaponItem> {
    return deepclone(MeleeWeapons())
  },

  get MiscEquipment(): ReadonlyArray<EquipmentItem> {
    return deepclone(MiscEquipment())
  },

  get MissileWeapons(): ReadonlyArray<MissileWeaponItem> {
    return deepclone(MissileWeapons())
  },
}

export default Equipment
