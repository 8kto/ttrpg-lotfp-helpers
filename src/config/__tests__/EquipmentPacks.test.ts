import { EquipmentPacks } from '@/config/EquipmentPacks'

describe('EquipmentPacks', () => {
  describe('snapshots match', () => {
    it('whole set', () => {
      expect(EquipmentPacks).toMatchSnapshot()
    })
  })
})
