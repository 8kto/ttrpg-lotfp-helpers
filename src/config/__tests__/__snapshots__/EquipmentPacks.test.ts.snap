// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EquipmentPacks snapshots match whole set 1`] = `
{
  "Adventurer": {
    "items": [
      [
        "Book, Blank",
        1,
      ],
      [
        "Candle",
        1,
      ],
      [
        "Chalk",
        1,
      ],
      [
        "Crowbar",
        1,
      ],
      [
        "Flask of Lamp Oil",
        3,
      ],
      [
        "Ink",
        1,
      ],
      [
        "Instrument",
        1,
      ],
      [
        "Lantern",
        1,
      ],
      [
        "Lock",
        1,
      ],
      [
        "Mallet",
        1,
      ],
      [
        "Mirror, Steel",
        1,
      ],
      [
        "Nails",
        20,
      ],
      [
        "Pole, 10'",
        1,
      ],
      [
        "Rope, 50'",
        1,
      ],
      [
        "Soap",
        1,
      ],
      [
        "Spike, Wooden",
        1,
      ],
      [
        "Tinderbox",
        1,
      ],
      [
        "Torch",
        3,
      ],
      [
        "Whistle",
        1,
      ],
    ],
    "name": "Adventurer",
  },
  "Basic": {
    "items": [
      [
        "Bedroll",
        1,
      ],
      [
        "Rations, Standard/Day",
        3,
      ],
      [
        "Backpack",
        1,
      ],
      [
        "Pouch",
        1,
      ],
      [
        "Sack",
        1,
      ],
      [
        "Tinderbox",
        1,
      ],
    ],
    "name": "Basic",
  },
  "Cleric1": {
    "items": [
      [
        "Leather",
        1,
      ],
      [
        "Shield",
        1,
      ],
      [
        "Holy Symbol, Steel",
        1,
      ],
      [
        "Mace",
        1,
      ],
      [
        "Holy Water",
        1,
      ],
    ],
    "name": "Cleric 1",
  },
  "Cleric2": {
    "items": [
      [
        "Leather",
        1,
      ],
      [
        "Holy Symbol, Steel",
        1,
      ],
      [
        "Maul",
        1,
      ],
      [
        "Holy Water",
        1,
      ],
      [
        "Sling",
        1,
      ],
      [
        "Sling bullets (10)",
        1,
      ],
    ],
    "name": "Cleric 2",
  },
  "Fighter1": {
    "items": [
      [
        "Chain",
        1,
      ],
      [
        "Shield",
        1,
      ],
      [
        "Two-handed sword",
        1,
      ],
    ],
    "name": "Fighter 1",
  },
  "Fighter2": {
    "items": [
      [
        "Chain",
        1,
      ],
      [
        "Polearm",
        1,
      ],
    ],
    "name": "Fighter 2",
  },
  "Fighter3": {
    "items": [
      [
        "Leather",
        1,
      ],
      [
        "Two-handed sword",
        1,
      ],
      [
        "Bow, Short",
        1,
      ],
      [
        "Arrows, quiver (20)",
        1,
      ],
    ],
    "name": "Fighter 3",
  },
  "MagicUser": {
    "items": [
      [
        "Leather",
        1,
      ],
      [
        "Book, Spell (Blank)",
        1,
      ],
      [
        "Ink",
        1,
      ],
      [
        "Dagger",
        2,
      ],
      [
        "Staff",
        1,
      ],
    ],
    "name": "Magic user",
  },
  "Specialist": {
    "items": [
      [
        "Leather",
        1,
      ],
      [
        "Specialistâ€™s Tools",
        1,
      ],
      [
        "Short sword",
        1,
      ],
      [
        "Dagger",
        2,
      ],
      [
        "Rope, 50'",
        1,
      ],
    ],
    "name": "Specialist",
  },
}
`;
